Add production checklist for ScalarDL (#215)
scalar-labs/scalar-kubernetes: docs/ProductionChecklistForScalarDLAuditor.md

  subgraph "ScalarDL Auditor mode"
    subgraph "Private network"
      subgraph "Kubernetes cluster for Ledger"
        B-1[ScalarDL Ledger]
      end
    end
    subgraph "Private network"
      subgraph "Kubernetes cluster for Auditor"
        C-1[ScalarDL Auditor]
      end
    end
  end
```

#### Not recommended for production environments (for testing purposes only)

```mermaid
graph LR
  subgraph "Kubernetes cluster"
    direction LR
    A-1[ScalarDL Ledger]
    A-2[ScalarDL Auditor]
  end
```

### Connect ScalarDL Ledger and ScalarDL Auditor

For ScalarDL Auditor mode to work properly, you must allow the connection between ScalarDL Ledger and ScalarDL Auditor.

```mermaid
graph LR
  subgraph "Kubernetes cluster for Ledger"
    A-1[ScalarDL Ledger]
  end
  subgraph "Kubernetes cluster for Auditor"
    B-1[ScalarDL Auditor]
  end
  A-1 --- B-1
```

ScalarDL uses the following ports for the connections between ScalarDL Ledger and ScalarDL Auditor. You must allow these connections between ScalarDL Ledger and ScalarDL Auditor. 

* ScalarDL Ledger
  * 50051/TCP
  * 50052/TCP
* ScalarDL Auditor
  * 40051/TCP
  * 40052/TCP

### Secret key and certification file requirements

> You must make sure that your secret key and certification file that you register to ScalarDL Ledger and ScalaDL Auditor match the following ScalarDL requirements.

Author: I see. I agree with you. We should clarify `PKI` or `HMAC`.
Also, indeed, we specify private keys and certificates in PEM format.
I will apply your suggestion. Thank you!

# Production checklist for ScalarDL Auditor

This checklist provides recommendations when deploying ScalarDL Auditor in a production environment.

## Before you begin

In this checklist, we assume that you are deploying ScalarDL Auditor on a managed Kubernetes cluster, which is recommended.

## Production checklist: ScalarDL Auditor

The following is a checklist of recommendations when setting up ScalarDL Auditor in a production environment.

> ### Number of pods and Kubernetes worker nodes

Author: @feeblefakie @josh-wong  
Thank you for your review!

> Pods and worker nodes (unify the current two sections, or having two sections like `Pods` and `Worker nodes`)

In this section (current `Number of pods and Kubernetes worker nodes` section), I mention both `Pods` and `Worker nodes` as follows.

```console
you should use at least three worker nodes and deploy at least three pods spread across the worker nodes
```

And, I want to tell these two things in one section to describe about availability.

Also, what I want to tell in this section is availability. So, I think we can update this section title to `Availability`.

In addition, in the current `Worker node specifications`, what I want to tell is resource estimation. So, I think we can update that section title to `Resources`.

With these in mind, I re-considered section titles as follows.

* Checklist for ScalarDL Ledger
  * Production checklist: ScalarDL Ledger
    * Availability (old section name is `Number of pods and Kubernetes worker nodes`)
    * Resources (old section name is `Worker node specifications`)
    * Network
    * Monitoring and logging
    * Backup and restore
  * Production checklist: Client applications that access ScalarDL Ledger
    * Client application deployment (old section name is `Do not deploy your application on the same Kubernetes cluster as ScalarDL Ledger`)
    * Contract and Function (old section name is `Check if your contract and function follow guidelines`)
    * Contract versioning (old section name is `Versioning contracts`)
    * Contract limitations (old section name is `Use the same binary name, package name, and class name when you register the contract`)
    * Private key and certificate (old section name is `Private key and certificate requirements`)
    * Exception handling
* Checklist for ScalarDL Auditor
  * Production checklist: ScalarDL Auditor
    * Availability (old section name is `Number of pods and Kubernetes worker nodes`)
    * Resources (old section name is `Worker node specifications`)
    * Network
    * Monitoring and logging
    * Backup and restore
    * ScalarDL Auditor deployment (old section name is `Do not deploy ScalarDL Auditor on the same Kubernetes cluster as ScalarDL Ledger`)
    * Connection between ScalarDL Ledger and ScalarDL Auditor (old section name is `Connect ScalarDL Ledger and ScalarDL Auditor`)
    * Private key and certificate (old section name is `Private key and certificate requirements`)
  * Production checklist: Client applications that access ScalarDL Auditor
    * Client application deployment (old section name is `Do not deploy your application on the same Kubernetes cluster as ScalarDL`)
    * Client application checklist (old section name is `Checklist for the client applications that access ScalarDL Ledger`)

What do you think?
(For consistency, we have to update the checklist of ScalarDB, but I will do it in another PR after this PR is merged.)

> ### Number of pods and Kubernetes worker nodes

Reviewer: Sorry, I didn't see this comment until @kota2and3kan's reply😔

Yes, after coming back to this, I can see how we can improve the headings. I agree with the changes @kota2and3kan suggested, with some minor changes for clarity:

**Checklist for ScalarDL Ledger**

- For `Availability`, I feel that `Pods and worker nodes` might be more accurate for this section. But maybe I'm not understanding this section clearly. 
  - If `Availability` is indeed a better phrase for what we're explaining in this part, how about `Cluster availability` to provide a little more context?
- `Contract and Function` -> `Contract and function`

**Checklist for ScalarDL Auditor**

- For `Availability`, I feel that `Pods and worker nodes` might be more accurate for this section. But maybe I'm not understanding this section clearly. 
  - If `Availability` is indeed a better phrase for what we're explaining in this part, how about `Cluster availability` to provide a little more context?

> ### Number of pods and Kubernetes worker nodes

Author: @josh-wong  
Thank you for your comment!

Strictly, the `availability` in this section means two things:

1. Kubernetes cluster's availability.
2. Scalar products' (i.e., ScalarDL's) availability.

Regarding `1.`, `Cluster availability` makes sense for me because it can indicate `Kubernetes cluster availability`.

However, regarding `2.`, I think `Cluster availability` is not good. This is because we don't call `ScalarDL` as `ScalarDL cluster`.

So, how about `Availability of Kubernetes and ScalarDL`?

---

As another suggestion, I think the simple title `ScalarDL availability` is also good.

The `Kubernetes availability` is a base of the `ScalarDL availability`. In other words, to keep `ScalarDL availability` is our final goal. To achieve that, we have to keep `Kubernetes availability` as a platform availability.

So, I think we can focus on our final goal `ScalarDL availability` in the section title.

I think `ScalarDL availability` is better. 

What do you think?

> ### Number of pods and Kubernetes worker nodes

Reviewer: @kota2and3kan Ah, thank you! I understand now. I was only considering `1.`

> The `Kubernetes availability` is a base of the `ScalarDL availability`. In other words, to keep `ScalarDL availability` is our final goal. To achieve that, we have to keep `Kubernetes availability` as a platform availability.
> 
> So, I think we can focus on our final goal `ScalarDL availability` in the section title.

I agree that `ScalarDL availability` would be better now that I understand we're concerned with ensuring that ScalarDL, which in essence means Kubernetes the platform, can withstand failures. I'm OK with that wording👍

> ### Number of pods and Kubernetes worker nodes

Author: @feeblefakie @josh-wong  
I updated the headings in 3c3b71594d5d5f8df28a738dd348d07a10a2e500 based on the discussion in this thread!
Please take a look when you have time!

> ### Number of pods and Kubernetes worker nodes

Reviewer 2: I feel `Availability` and `ScalarDL Availability` sound a bit inconsistent with the others to me.
(I feel, if we name it Availability, then the others should be like `Observability` for monitoring and logging and `Durability` for backup and restore.)

But, I'm fine with it for now, and we can rethink about it later on.

> ### Number of pods and Kubernetes worker nodes

Author: @feeblefakie 
Thank you for your comment!

> I feel `Availability` and `ScalarDL Availability` sound a bit inconsistent with the others to me.
(I feel, if we name it Availability, then the others should be like `Observability` for monitoring and logging and `Durability` for backup and restore.)

I see. Indeed, `Observability` looks more consistent title name.

> But, I'm fine with it for now, and we can rethink about it later on.

OK. Let's rethink/discuss it later!

To ensure that the Kubernetes cluster has high availability, you should use at least three worker nodes and deploy at least three pods spread across the worker nodes. You can see the [sample configurations](../conf/scalardl-audit-custom-values.yaml) of `podAntiAffinity` for making three pods spread across the worker nodes.

{% capture notice--info %}
**Note**

If you place the worker nodes in different availability zones (AZs), you can withstand an AZ failure.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

### Worker node specifications

From the perspective of commercial licenses, resources for one pod running ScalarDL Auditor are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDL Auditor pods exist on the worker nodes.

> In other words, the following components could run on one worker node:

Reviewer: Does the wording for this depend on how we word a similar sentence in a different PR?: [related comment](https://github.com/scalar-labs/scalar-kubernetes/pull/213/files#r1382880277)

Author: Yes! I will update this sentence after the discussion in another PR is finished!

Author: Fixed in 259da4d754e7b551d0ade62bef962476cdc79b6c.

### Network

You should create the Kubernetes cluster on a private network since ScalarDL Auditor does not provide any services to users directly via internet access. We recommend accessing ScalarDL Auditor via a private network from your applications.

### Monitoring and logging

You should monitor the deployed components and collect their logs. For details, see [Monitoring Scalar products on a Kubernetes cluster](./K8sMonitorGuide.md) and [Collecting logs from Scalar products on a Kubernetes cluster](./K8sLogCollectionGuide.md).

### Backup and restore

You should enable the automatic backup feature and point-in-time recovery (PITR) feature in the backend database. For details, see [Set up a database for ScalarDB/ScalarDL deployment](./SetupDatabase.md).

### Do not deploy ScalarDL Auditor on the same Kubernetes cluster as ScalarDL Ledger

> For Byzantine fault detection in ScalarDL to work properly, do not deploy ScalarDL Auditor pods on the same Kubernetes clusters as the ScalarDL Ledger deployment. Please deploy ScalarDL Auditor pods on an environment other than the administrative domain (other than the Kubernetes cluster) for ScalarDL Ledger deployment.
> 

Reviewer: ```suggestion
For Byzantine fault detection in ScalarDL to work properly, do not deploy ScalarDL Auditor pods on the same Kubernetes clusters as the ScalarDL Ledger deployment. Instead, you must deploy ScalarDL Auditor pods in an environment other than the administrative domain (other than the Kubernetes cluster) for the ScalarDL Ledger deployment.
```

#### Required for production environments

```mermaid
graph LR

>   subgraph "ScalarDL Auditor mode"

Reviewer 2: ```suggestion
  subgraph "ScalarDL"
```
This subgraph includes Ledger and Auditor, so naming it `Auditor mode` is probably misleading.

Author: Ah, I see. You are right. It's a bit confusing.
I will apply your suggestion! Thank you!

    subgraph "Administrative domain 1"
      subgraph "Kubernetes cluster for Ledger"
        B-1[ScalarDL Ledger]
      end
    end
    subgraph "Administrative domain 2"
      subgraph "Kubernetes cluster for Auditor"
        C-1[ScalarDL Auditor]
      end
    end
  end
```

#### Not recommended for production environments (for testing purposes only)

```mermaid
graph LR
  subgraph "Kubernetes cluster"
    direction LR
    A-1[ScalarDL Ledger]
    A-2[ScalarDL Auditor]
  end

> ```
> 

Reviewer: ````suggestion
```
````

### Connect ScalarDL Ledger and ScalarDL Auditor

For ScalarDL Auditor mode to work properly, you must allow the connection between ScalarDL Ledger and ScalarDL Auditor.

```mermaid
graph LR
  subgraph "Kubernetes cluster for Ledger"
    A-1[ScalarDL Ledger]
  end
  subgraph "Kubernetes cluster for Auditor"
    B-1[ScalarDL Auditor]
  end
  A-1 --- B-1
```

> ScalarDL uses the following ports for the connections between ScalarDL Ledger and ScalarDL Auditor. You must allow these connections between ScalarDL Ledger and ScalarDL Auditor. 

Reviewer: ```suggestion
ScalarDL uses the following ports for the connections between ScalarDL Ledger and ScalarDL Auditor. You must allow these connections between ScalarDL Ledger and ScalarDL Auditor:
```

* ScalarDL Ledger
  * 50051/TCP
  * 50052/TCP
* ScalarDL Auditor
  * 40051/TCP
  * 40052/TCP

### Private key and certificate requirements

> When you use PKI for authentication, you must make sure that private keys and certificates that you register to ScalarDL Ledger and ScalaDL Auditor match the following requirements.

Reviewer: ```suggestion
When you use PKI for authentication, you must make sure that private keys and certificates that you register to ScalarDL Ledger and ScalaDL Auditor match the following requirements:
```

```console
Algorithm       : ECDSA
Hash function   : SHA256
Curve parameter : P-256
```

For details, see [How to get a certificate](https://github.com/scalar-labs/scalardl/blob/master/docs/ca/caclient-getting-started.md).

## Production checklist: Client applications that access ScalarDL Auditor

The following is a checklist of recommendations when setting up a client application that accesses ScalarDL Auditor in a production environment.

### Do not deploy your application on the same Kubernetes cluster as ScalarDL

> For Byzantine fault detection in ScalarDL to work properly, do not deploy your application pods on the same Kubernetes clusters as the ScalarDL deployment. Please deploy your application on an environment other than the administrative domain (other than the Kubernetes cluster) for ScalarDL deployment.

Reviewer: ```suggestion
For Byzantine fault detection in ScalarDL to work properly, do not deploy your application pods on the same Kubernetes clusters as the ScalarDL deployment. Instead, you must deploy your application in an environment other than the administrative domain (other than the Kubernetes cluster) for the ScalarDL deployment.
```

#### Required for production environments

```mermaid
graph LR
  subgraph "Administrative domain 1"
    subgraph "Another environment"
      A-1[User application]
    end
  end
  subgraph "ScalarDL"
    subgraph "Administrative domain 2"
      subgraph "Kubernetes cluster for Ledger"
        B-1[ScalarDL Ledger]
      end
    end
    subgraph "Administrative domain 3"
      subgraph "Kubernetes cluster for Auditor"
        C-1[ScalarDL Auditor]
      end
    end
  end
  A-1 --> B-1
  A-1 --> C-1
```

#### Not recommended for production environments (for testing purposes only)

```mermaid
graph LR
  subgraph "Kubernetes cluster"
    direction LR
    A-1[User application]
    A-2[ScalarDL Ledger]
    A-3[ScalarDL Auditor]
  end
  A-1 --> A-2
  A-1 --> A-3
```

### Checklist for the client applications that access ScalarDL Ledger

> You also must make sure that you satisfy the [Production checklist: Client applications that access ScalarDL Ledger](./ProductionChecklistForScalarDLLedger.md#production-checklist-client-applications-that-access-scalardl-ledger).

Reviewer: ```suggestion
You also must make sure that you satisfy the [Production checklist: Client applications that access ScalarDL Ledger](./ProductionChecklistForScalarDLLedger.md#production-checklist-client-applications-that-access-scalardl-ledger).

```
